File: src/main/java/com/github/kindrat/cassandra/client/ui/window/editor/main/BackgroundTaskMonitor.java
package com.github.kindrat.cassandra.client.ui.window.editor.main;

import com.github.kindrat.cassandra.client.properties.UIProperties;
import com.github.kindrat.cassandra.client.service.tasks.TaskStatistics;
import com.github.kindrat.cassandra.client.util.UIUtil;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.layout.Pane;
import lombok.extern.slf4j.Slf4j;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

import static com.github.nginate.commons.lang.NStrings.format;

@Slf4j
public class BackgroundTaskMonitor extends Pane {
    private static final String TEMPLATE = "{} ⏲";
    private final AtomicInteger counter = new AtomicInteger();
    private final Map<UUID, TaskStatistics> taskStatistics = new HashMap<>();

    private final Label label;

    public BackgroundTaskMonitor(UIProperties properties) {
        label = new Label(format(TEMPLATE, counter.get()));
        label.alignmentProperty().setValue(Pos.BASELINE_CENTER);
        UIUtil.setWidth(this, 20);
        getChildren().add(label);
    }

    public void addTask(UUID id) {
        taskStatistics.put(id, null);
        label.setText(format(TEMPLATE, counter.incrementAndGet()));
    }

    public void update(UUID id, TaskStatistics statistics) {
        taskStatistics.put(id, statistics);
    }

    public void remove(UUID id) {
        taskStatistics.remove(id);
        label.setText(format(TEMPLATE, counter.decrementAndGet()));
    }

    public void showTasks() {
        log.debug("!!!!!!!!!!");
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/window/editor/main/BackgroundTaskMonitor.kt
package com.github.kindrat.cassandra.client.ui.window.editor.main

import com.github.kindrat.cassandra.client.properties.UIProperties
import com.github.kindrat.cassandra.client.service.tasks.TaskStatistics
import com.github.kindrat.cassandra.client.util.UIUtil.setWidth
import javafx.geometry.Pos
import javafx.scene.control.Label
import javafx.scene.layout.Pane
import mu.KotlinLogging
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

private val logger = KotlinLogging.logger {}

class BackgroundTaskMonitor(properties: UIProperties) : Pane() {
    private val counter = AtomicInteger()
    private val taskStatistics: MutableMap<UUID, TaskStatistics?> = HashMap()
    private val label: Label

    fun addTask(id: UUID) {
        taskStatistics[id] = null
        label.text = "${counter.incrementAndGet()} ⏲"
    }

    fun update(id: UUID, statistics: TaskStatistics?) {
        taskStatistics[id] = statistics
    }

    fun remove(id: UUID?) {
        taskStatistics.remove(id)
        label.text = "${counter.decrementAndGet()} ⏲"
    }

    fun showTasks() {
        logger.debug("!!!!!!!!!!")
    }

    companion object {
        private const val TEMPLATE = "{} ⏲"
    }

    init {
        label = Label("${counter.get()} ⏲")
        label.alignmentProperty().value = Pos.BASELINE_CENTER
        setWidth(this, 20.0)
        children.add(label)
    }
}